pkg/controllers/dependenciesdistributor/dependencies_distributor.go

```diff
@@ -47,100 +53,46 @@ const (
-	DynamicClient dynamic.Interface
-
+	DynamicClient       dynamic.Interface
 	InformerManager     genericmanager.SingleClusterInformerManager
 	EventHandler        cache.ResourceEventHandler
 	EventRecorder       record.EventRecorder
 	Processor           util.AsyncWorker
 	RESTMapper          meta.RESTMapper
 	ResourceInterpreter resourceinterpreter.ResourceInterpreter
+	RateLimiterOptions  ratelimiterflag.Options
 
 	// bindingReconcileWorker maintains a rate limited queue which used to store ResourceBinding's key and
 	// a reconcile function to consume the items in queue.
 	bindingReconcileWorker util.AsyncWorker
-	resourceBindingLister  cache.GenericLister
-
-	stopCh <-chan struct{}
-}
-
-// Start runs the distributor, never stop until stopCh closed.
-func (d *DependenciesDistributor) Start(ctx context.Context) error {
-	klog.Infof("Starting dependencies distributor.")
-	d.stopCh = ctx.Done()
-
-	bindingWorkerOptions := util.Options{
-		Name:          "resourceBinding reconciler",
-		KeyFunc:       detector.ClusterWideKeyFunc,
-		ReconcileFunc: d.ReconcileResourceBinding,
-	}
-	// setup binding reconcile worker
-	d.bindingReconcileWorker = util.NewAsyncWorker(bindingWorkerOptions)
-	d.bindingReconcileWorker.Run(2, d.stopCh)
-
-	// watch and enqueue ResourceBinding changes.
-	resourceBindingGVR := schema.GroupVersionResource{
-		Group:    workv1alpha2.GroupVersion.Group,
-		Version:  workv1alpha2.GroupVersion.Version,
-		Resource: "resourcebindings",
-	}
-
-	bindingHandler := fedinformer.NewHandlerOnEvents(nil, d.OnResourceBindingUpdate, d.OnResourceBindingDelete)
-	d.InformerManager.ForResource(resourceBindingGVR, bindingHandler)
-	d.resourceBindingLister = d.InformerManager.Lister(resourceBindingGVR)
-	resourceWorkerOptions := util.Options{
-		Name:          "resource detector",
-		KeyFunc:       detector.ClusterWideKeyFunc,
-		ReconcileFunc: d.Reconcile,
-	}
-	d.EventHandler = fedinformer.NewHandlerOnEvents(d.OnAdd, d.OnUpdate, d.OnDelete)
-	d.Processor = util.NewAsyncWorker(resourceWorkerOptions)
-	d.Processor.Run(2, d.stopCh)
-	go d.discoverResources(30 * time.Second)
-
-	<-d.stopCh
-
-	klog.Infof("Stopped as stopCh closed.")
-	return nil
+	stopCh                 <-chan struct{}
 }
 
 // Check if our DependenciesDistributor implements necessary interfaces
 var _ manager.Runnable = &DependenciesDistributor{}
 var _ manager.LeaderElectionRunnable = &DependenciesDistributor{}
 
+// Reconcile performs a full reconciliation for the object referred to by the Request.
+// The Controller will requeue the Request to be processed again if an error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (d *DependenciesDistributor) Reconcile(_ context.Context, request reconcile.Request) (reconcile.Result, error) {
+	d.bindingReconcileWorker.Add(keys.ClusterWideKey{Namespace: request.Namespace, Name: request.Name})
+	return reconcile.Result{}, nil
+}
```

Comment:
I think we can have a formal controller for `dependencies-distributor`, but I don't think that's the way to do it.

The main job of the controller is responsible to reconcile RB changes and sync schedule results to dependencies( RBs referencing dependencies like ConfigMap/Secret). That means we don't need the `bindingReconcileWorker` anymore.

Reply:
> The main job of the controller is responsible to reconcile RB changes and sync schedule results to dependencies( RBs referencing dependencies like ConfigMap/Secret). That means we don't need the bindingReconcileWorker anymore.


Yes, I also considered removing `bindingReconcileWorker`, but when I saw L443 `d.bindingReconcileWorker.Run(2, d.stopCh)`, I chose not to remove it because I understand that `bindingReconcileWorker` can be more efficient than the `formal controller`.